Text/LaTeX/Base/Syntax/WithParm.hs

```diff
@@ -0,0 +1,347 @@
+{-# LANGUAGE FlexibleInstances, CPP, DeriveDataTypeable, DeriveGeneric, DeriveFunctor #-}
+
+-- | LaTeX syntax description in the definition of the 'LaTeX' datatype.
+--   If you want to add new commands or environments not defined in
+--   the library, import this module and use 'LaTeX' data constructors.
+module Text.LaTeX.Base.Syntax.WithParm
+ ( -- * @LaTeX@ datatype
+   Measure (..)
+ , MathType (..)
+ , LaTeX (..)
+ , TeXArg (..)
+ , (<>), between
+   -- * Escaping reserved characters
+ , protectString
+ , protectText
+   -- * Syntax analysis
+ , matchCommand
+ , lookForCommand
+ , matchEnv
+ , lookForEnv
+ , texmap 
+ , texmapM
+   -- ** Utils
+ , getBody
+ , getPreamble
+   ) where
+
+import Data.Text (Text,pack)
+import qualified Data.Text
+import qualified Data.Semigroup as Semigroup
+import Data.String
+import Control.Applicative
+import Control.Monad (replicateM)
+import Data.Functor.Identity (runIdentity)
+import Data.Data (Data)
+import Data.Typeable
+import Test.QuickCheck
+import Data.Hashable
+import GHC.Generics (Generic)
+#if !MIN_VERSION_base(4,11,0)
+import Data.Monoid
+#endif
+
+-- | Measure units defined in LaTeX. Use 'CustomMeasure' to use commands like 'textwidth'.
+--   For instance:
+--
+-- > rule Nothing (CustomMeasure linewidth) (Pt 2)
+--
+-- This will create a black box (see 'rule') as wide as the text and two points tall.
+--
+data Measure a =
+   Pt Double -- ^ A point is 1/72.27 inch, that means about 0.0138 inch or 0.3515 mm.
+ | Mm Double -- ^ Millimeter.
+ | Cm Double -- ^ Centimeter.
+ | In Double -- ^ Inch.
+ | Ex Double -- ^ The height of an \"x\" in the current font.
+ | Em Double -- ^ The width of an \"M\" in the current font.
+ | CustomMeasure (LaTeX a) -- ^ You can introduce a 'LaTeX' expression as a measure.
+   deriving (Data, Eq, Generic, Show, Typeable, Functor)
+
+-- | Different types of syntax for mathematical expressions.
+data MathType = Parentheses | Square | Dollar | DoubleDollar | LHSInline
+  deriving (Data, Eq, Generic, Show, Typeable)
+
+-- | Type of @LaTeX@ blocks. The @a@ type variable can be used for
+-- a number of things but mainly stores source location.
+data LaTeX a =
```

Comment:
Does it really make sense to stuff the `a` parameter in every single constructor? That seems excessive.

In fact, how about going about it the opposite way: we could introduce just _a single, new_ constructor, that's responsible for nothing else but position information etc., and is included in the AST through `TeXSeq`. The “no positions mode” would be `LaTeX Void` instead of `LaTeX ()`.

Reply:
I agree it seems a little excessive. I'm unsure about the single constructor.

Lets say we add a single constructor, call it `TeXLoc`, now
Imagine that we are using locations for the sake of error reporting, and we are currently processing a TeXEnv constructor: if we wish to raise an error and indicate to the user where the error came from in the source file, we'd have to find the closest `TeXLoc` and hope its accurate enough.

I'll think this through a little more.

Reply:
@leftaroundabout I have removed the src location from most constructors; now only `TeXRawL`, `TeXCommSL`, `TeXCommL`, `TeXEnvL` and `TeXMathL` carry source location (as indicated by their `L` suffix). The `TeXEnvL` carries two locations corresponding to  where the respective `\begin` and `\end` are.

I also changed the parser to add such annotations.